### PR TITLE
Switch from html2text to soundasleep/html2text

### PR DESCRIPTION
**Pull request by larssandergreen**
Overview
----------------------------------------
[html2text/html2text](https://github.com/mtibben/html2text) seems to do weird things including turning bold to SHOUTY, changing `<td>` to newlines and some kind of indentation for tables within tables. [soundasleep/html2text](https://github.com/soundasleep/html2text/) seems a little more sane, it ignores bold, changes `<td>` to tabs, and ignores tables within tables. The message templates seem a little more reasonable this way.

This would be in service of #27688.

So let's see what breaks with this!

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@larssandergreen cool - will be interesting!

**Comment by larssandergreen** [comment]
For example:
html2text
```

 

Dear Jed,

Contact the Development Department if you need to make any changes to your
registration.

 		 EVENT INFORMATION AND LOCATION 

 		 Fall Fundraiser Dinner
August 31st, 2023 5:00 PM-November 2nd, 2024 

 		 14S El Camino Way E
Collinsville, Connecticut 6022

 		 Event Contacts: 

 		 Phone 
 		 204 222-1000 

 		 Email 
 		 development@example.org 

 		 Download iCalendar entry for this event. [1] 

 		 Add event to Google Calendar [2] 

 		 REGISTERED EMAIL 

 		 dimitrov.o.jed@fishmail.org 

 		 DINNER CONTRIBUTION 

 		ITEM
 		QTY
 		EACH
 		SUBTOTAL
 		TAX RATE
 		TAX AMOUNT
 		TOTAL

 		Festival Fee - Bass
 		1
 		$25.00
 		$25.00
 		0.00%
 		$0.00
 		 $25.00 

 		Something
 		3
 		$2.00
 		$6.00
 		7.00%
 		$0.42
 		 $6.42 

 		Checkboxes - One
 		1
 		$1.00
 		$1.00
 		7.00%
 		$0.07
 		 $1.07 

 		 Amount Before Tax: 
 		 $32.00 

 		No Sales Tax
 		$0.00

 		Sales Tax 7.00%
 		$0.49

 		 Total Tax Amount 
 		 $0.49 

 		 Total Amount 
 		 $32.49 

 		 Registration Date 
 		 September 30th, 2023 5:28 AM 

 		 Transaction Date 
 		 September 30th, 2023 5:28 AM 

 		 Financial Type 
 		 Event Fee 

 		 Paid By 
 		 Check 

 		 FOOD PREFERENCE 

 		 Soup Selection 

 

Links:
------
[1] http://drupal-test.localhost/civicrm/event/ical?reset=1&id=1
[2] http://drupal-test.localhost/civicrm/event/ical?gCalendar=1&reset=1&id=1

```

soundasleep:
```

Dear Jed,

Contact the Development Department if you need to make any changes to your registration.

Event Information and Location
Fall Fundraiser Dinner
August 31st, 2023 5:00 PM-November 2nd, 2024
14S El Camino Way E
Collinsville, Connecticut 6022
Event Contacts:
Phone	204 222-1000
Email	development@example.org
[Download iCalendar entry for this event.](http://drupal-test.localhost/civicrm/event/ical?reset=1&id=1)
[Add event to Google Calendar](http://drupal-test.localhost/civicrm/event/ical?gCalendar=1&reset=1&id=1)
Registered Email
dimitrov.o.jed@fishmail.org
Dinner Contribution

Item	Qty	Each	Subtotal	Tax Rate	Tax Amount	Total
Festival Fee - Bass	1	$25.00	$25.00	0.00%	$0.00	$25.00
Something	4	$2.00	$8.00	7.00%	$0.56	$8.56
Checkboxes - Two	1	$2.00	$2.00	7.00%	$0.14	$2.14
Amount Before Tax:	$35.00
No Sales Tax	$0.00
Sales Tax 7.00%	$0.70
Total Tax Amount	$0.70
Total Amount	$35.70
Registration Date	September 30th, 2023 5:26 AM
Transaction Date	September 30th, 2023 5:26 AM
Financial Type	Event Fee
Paid By	Check
Food Preference
Soup Selection
```

**Comment by eileenmcnaughton** [comment]
@larssandergreen that's impressive!

**Comment by konadave** [comment]
html2text also likes to randomly change `{ts}Whatever{/ts}` into `{TS}WHATEVER{/TS}` which causes Smarty to not recognize (at least v2) the `{ts}` tag.

**Comment by larssandergreen** [comment]
So it turns out we'll have to use the [ignore XML parsing errors option](https://github.com/soundasleep/html2text#options) for this, as it parses the DOMDocument as XML which does not allow literal ampersands, even in URLs. That seems OK to me, we're just converting whatever HTML we have to text, if the HTML is a bit funky, we still want to convert it to text.

The remainder of the errors look like cases where the output has changed (mostly inline URLs instead of the footnote format), so those are easy fixes.

If we can get some agreement that this is a good idea in general, then I'll get the remaining errors sorted.

**Comment by eileenmcnaughton** [comment]
@larssandergreen yes - if you can get this to pass I think we should merge it - I'm sold

File: ext/flexmailer/tests/phpunit/Civi/FlexMailer/ClickTracker/TextClickTrackerTest.php [hunk 1]
```diff
@@ -45,19 +45,14 @@ public function getHrefExamples() {
       '<p><a href=\'tracking(https://sub.example.com/foo.php?whiz=%2Fbang%2F&pie[fruit]=apple)\' rel=\'nofollow\'>Foo</a></p>',
     ];
     $exs[] = [
-      // Messy looking URL, designed to trip-up quote handling
```
Comment (larssandergreen) [inline ext/flexmailer/tests/phpunit/Civi/FlexMailer/ClickTracker/TextClickTrackerTest.php:48]: Because we previously converted links into [Text] plus a footnote with the URL, all we were testing here was of `[Foo] === [Foo]` over and over. Now that the URL is inline in parentheses, these tests actually do something (and so they can fail).

The two javascript:alert ones shouldn't have tracking prepended, because the regex we're using only matches urls with http in them, so these don't count as urls and don't get the tracking. Probably not so valuable, but we can leave them in here just to make sure the tracking isn't added.

I've removed the one with a newline in the url. I don't think that's a reasonable thing to test or there is any need for us to support a url with a newline in it (url-encoded, sure, but a literal `/n`, no).

**Comment by eileenmcnaughton** [comment]
This looks good to me but I'd like @seamuslee001 or @totten or @demeritcowboy to run their eyes over the url changes you commented to confirm there isn't anything that could sneak through security-wise

**Comment by demeritcowboy** [comment]
On the urls, there is [this commit](https://github.com/civicrm/org.civicrm.flexmailer/commit/25ab3efaa3088988b9ea1d778842d4ad6bb135cf), which was about how the custom clicktracker code handled urls, and how it handled quotes. I think for quotes handling there isn't a change in test coverage.

While "javascript" urls are always scary I'm not sure they have meaning in text/plain emails because I'd be amazed if those are runnable/clickable. It's more about the quotes I think, and is probably more relevant to the html version where this was copied from originally.

I haven't checked yet to see if/when these stopped having tracking applied, but that does sound kind of familiar, so might be in the last few years.

Someday, at a taking-a-step-back-level, it seems worth considering replacing all text versions with the stock phrase "There is no text version, silly." Just to satisfy the RFC/spam-checkers, but no human will ever see it. And I'd be amazed if anyone is relying on the **automatic translation** of civi emails into text for B2B automated transfer of data or processing.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy I hadn't thought about the spam checkers - do you think they still look at that?

**Comment by demeritcowboy** [comment]
I don't know how high they rank it or exactly what they look for. It's required by the RFC to have _something_ text/plain, but I doubt it's a big factor for spam.

**Comment by seamuslee001** [comment]
I also wondered about if having the links in the footer was a requirement under the RFC or not

**Comment by demeritcowboy** [comment]
I doubt it. That's probably copied from an old thing called "books", where you had numbered "footnotes", and everyone just understood it at the time.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy hmm OK - but my take away from your comment is that you don't see anything as blocking us merging this & then moving towards removing the text versions of the templates in favour of this library converting the html

Of course that assumes there is nothing in @seamuslee001's concern re the RFC

**Comment by demeritcowboy** [comment]
I don't have any theoretical concerns. I haven't run it.

I did also check into a previous issue with wordpress mailpoet which uses this library, but it shouldn't conflict.

**Comment by totten** [comment]
(1) __Squishy/Feely/Peronal/Bikesheddy Reactions__:  On the whole, in [this example](https://github.com/civicrm/civicrm-core/pull/27694#issuecomment-1741689123), the new `html2text` feels slightly better. The tabular-table is the big draw. I wish the new one generated more new-lines. I don't really care about `<strong>`/CAPS. For hyperlinks, both approaches (footnotes and markdown-ish) are weird to me.

FWIW, I think hard-written plain-text can do URLs better - you have to write your content in a way that puts the URLs into standalone paragraphs or definition-lists. (***Plain-content and rich-content should have different writing-styles.***) But I don't know if we actually follow-through on that potential, and I understand the audience that cares is tiny.

But on the whole, for the one example we have, the new converter looks better than the old one.

(2) __Examples__: Has anybody tried to view the range of messages in the different formatters?

(3) __Side-chat: Need for plain-text__: 

> (@demeritcowboy) I don't know how high they rank it (*text vs html*) or exactly what they look for. It's required by the RFC to have something text/plain, but I doubt it's a big factor for spam.

Yeah, this is also a squishy topic. FWIW, I did a quick Google search, and I still find recent-ish articles from email-vendors which suggest that the senders' choice of text-vs-html impacts delivery/opens. (I think a 2021 article from Hubspot said their open-rates were ~27% better on plain-text. Another vendor had 2023 article saying that HTML was more likely to land in a "Promotions" folder, but I don't remember their name, and they didn't give numbers)

If this was just about post-donation receipts, then I wouldn't worry about messages landing in "Promotions" folder. (*Personally, as a reader, I like having receipts in a side-folder.*) But for case-related notifications, delivery can impact the timeliness of the casework -- don't want those to be viewed as spammy. 

It's a bit annoying that we don't have the data-collection needed to assess this kind of thing for ourselves. Such is the lot of FOSS.

It's interesting to dig into one's personal mailbox to spot-check what the rest of the industry is doing. For most automated messages I saw, several made an effort to have plain-text. The quality/readability varied widely. A couple used a placeholder text like "Please view webpage XYZ". One was pure-HTML. In other words: *chaos*. 

(4) __Side-chat: General approaches to plain-text__

> (@demeritcowboy) Someday, at a taking-a-step-back-level, it seems worth considering replacing all text versions with the stock phrase "There is no text version, silly."

Yeah, I think it'd be interesting to compare the generic strategies for text-handling. You could potentially see that as a policy -- based on some setting or some extensions, eg

```
Administer > Mail Settings > Plain Text Email:

( ) Plain-text email is composed manually
( ) Plain-text email is downsampled from HTML email
       [ Pick converter ]
(x) Plain-text email is a place-holder message.
       [x] Customize placeholder:
           > This email was written in HTML. To view the full message, go to:
           > https://example.com/civicrm/mail?code=ASDFASDFASDF
( ) Plain-text is not sent
```

----

tldr, re:PR: The new converter looks a bit better than the old converter (*in the given example*), but I would suggest sampling more messages before concluding that it is better.

**Comment by eileenmcnaughton** [comment]
I think the new converter is passable on everything - there were a bunch of examples in the original test fails - whereas the old converter wasn't really passable on tables.

**Comment by eileenmcnaughton** [comment]
Chatted on this - & consensus is to merge
